### DISCUSSION TITLE
To off screen or not to off screen

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [cefsharp/CefSharp#4865] [Q&A]
That is the question.

I have a case where, it might be helpful to sometimes show the WPF browser `Window`, but most of the time, I think we will be content to maintain a browser presence _off screen_ as it were.

The challenge I am faced with, however, maybe why there is a separate [_off screen package_](https://www.nuget.org/packages/CefSharp.OffScreen) (?), is that I might like to support more of a hybrid approach. The window should be able to be made visible when necessary; but should still operate seamlessly off screen.

The inner workings, as far as I am aware, is that the browser is not rendering, and/or has not rendered, when the window is hidden. And more inward than that, the `BrowserCore` instance has not been initialized until that point. And which allows for things like context to be arranged and so forth.

So, the question for me is, is there a recommendation for this sort of hybrid approach?

**Comment by mwpowellhtx** [comment]
I am trying to arrange a browser control element as a tab item in a tab control. However, when I run a check whether I could sign in, it is hanging on the first control selector `await`. Which I am wondering if possibly being anywhere other than an active window, or otherwise caught up in the current window event loop, is not kosher.

Below, blurred for privacy, tab control and browser on the right, sign in controls on the left, obviously.

![image](https://github.com/cefsharp/CefSharp/assets/1824187/8d529785-7f19-4a54-b16e-9497c0149b68)

```csharp
await using var context = await browser.CreateDevToolsContextAsync();

var opts = DevToolsWaitForSelectorOptions;

// Hanging up here:
var a = await context.WaitForSelectorAsync(S.Password, opts);
//      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
var b = await context.WaitForSelectorAsync(S.Username, opts);
var c = await context.WaitForSelectorAsync(S.RememberMe, opts);
var d = await context.WaitForSelectorAsync(S.Submit, opts);

return !(a is null || b is null || c is null || d is null);
```

Options are nominal as far as I know:

```csharp
private static WaitForSelectorOptions DevToolsWaitForSelectorOptions => new() { Timeout = 1000, Visible = true };
```

I have validated the approach is more or less sound; but I wonder if I am encountering some off screen aspects. Rather how to persuade the await when the browser is another control in the same window.

**Reply by mwpowellhtx** [reply]
That I can tell, it seems to be with the `ICommand`, specifically, `CommunityToolkit.Mvvm.Input.RelayCommand` condition, does not seem to like `async` within its return area whatsoever. The action itself can be asynchronous, and it is _"fine"_. Just not the condition. But that's okay, I can work around that.

**Comment by amaitland** [comment]
> I have a case where, it might be helpful to sometimes show the WPF browser `Window`, but most of the time, I think we will be content to maintain a browser presence _off screen_ as it were.

The `OffScreen` and `WPF` implementations both use `Offscreen rendering` (every frame is rendered as a bitmap).

You can create the `WPF` browser before it's attached to the visual tree, there's a specific constructor overload for this.

See https://github.com/cefsharp/CefSharp/blob/cefsharp/126/CefSharp.Wpf/ChromiumWebBrowser.cs#L529

The example demos adding the browser when it's been created, you can just defer this until your ready to show it to the user.

You may also wish to change the framerate when the browser is not visible to reduce resource load.

**Reply by mwpowellhtx** [reply]
Good points all, good to know.

I was able to make it work on screen after all, and am not displeased with my ability to negotiate the DOM, puppeteer in strategic places, etc. Seems to be working pretty decently.

Great job!
